### PR TITLE
Implemented dependency pipeline for ADO

### PR DESCRIPTION
**Pull request by MrMCake**
# Change

- Implemented dependency pipeline for ADO
- Requries bugfix #844 

Tested reference
image

> **Note:** The private DNS Zone deployed failed due to an ADO unrelated issue with a lack of idempotency

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

### CODE REVIEW

**Review by rahalan — APPROVED** [review]

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep [hunk 1]
```diff
@@ -580,13 +580,13 @@ module virtualMachine_rbac '.bicep/nested_rbac.bicep' = [for (roleAssignment, in
 }]
 
 @description('The name of the VM.')
-output virtualMachineName string = virtualMachine.name
+output name string = virtualMachine.name
 
 @description('The Resource ID of the VM.')
-output virtualMachineResourceId string = virtualMachine.id
+output ResourceId string = virtualMachine.id
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/deploy.bicep:586]:
```suggestion
output resourceId string = virtualMachine.id
```

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep [hunk 2]
```diff
@@ -580,13 +580,13 @@ module virtualMachine_rbac '.bicep/nested_rbac.bicep' = [for (roleAssignment, in
...
 
 @description('The name of the Resource Group the VM was created in.')
-output virtualMachineResourceGroup string = resourceGroup().name
+output ResourceGroup string = resourceGroup().name
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/deploy.bicep:589]:
```suggestion
output resourceGroup string = resourceGroup().name
```

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 1 above)
Comment (eriqua) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep:586]: Please update the readme as well

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 2 above)
Comment (eriqua) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep:589]: Please update the readme as well

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep [hunk 3]
```diff
@@ -580,13 +580,13 @@ module virtualMachine_rbac '.bicep/nested_rbac.bicep' = [for (roleAssignment, in
 }]
 
 @description('The name of the VM.')
-output virtualMachineName string = virtualMachine.name
+output name string = virtualMachine.name
 
 @description('The Resource ID of the VM.')
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/deploy.bicep:585]:
```suggestion
@description('The resource ID of the VM.')
```

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 1 above)
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/deploy.bicep:586]:
```suggestion
output resourceId string = virtualMachine.id
```

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep [hunk 4]
```diff
@@ -580,13 +580,13 @@ module virtualMachine_rbac '.bicep/nested_rbac.bicep' = [for (roleAssignment, in
...
-output virtualMachineResourceId string = virtualMachine.id
+output ResourceId string = virtualMachine.id
 
 @description('The name of the Resource Group the VM was created in.')
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/deploy.bicep:588]:
```suggestion
@description('The name of the resource group the VM was created in.')
```

File: .github/workflows/platform.dependencies.yml [hunk 5]
```diff
@@ -38,8 +38,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        parameterFilePaths:
-          ['artifacts.parameters.json', 'validation.parameters.json']
+        parameterFilePaths: ['validation.parameters.json']
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:41]: Will the artifacts RG and container registry be created directly by the publish job?

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 1 above)
Comment (MrMCake) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep:586]: not supposed to be here. Will roll back

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 2 above)
Comment (MrMCake) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep:589]: not supposed to be here. Will roll back

File: .github/workflows/platform.dependencies.yml (hunk 5 above)
Comment (MrMCake) [reply .github/workflows/platform.dependencies.yml:41]: Yes, they always were - and from the feedback I received it rather raised questions than resolve them :/ I hence decided to remove it from the dependency pipeline. Happy for feedback if we want to keep it regardless

File: .github/workflows/platform.dependencies.yml [hunk 6]
```diff
@@ -650,33 +626,56 @@ jobs:
...
+              @{ name = 'administratorLoginPassword'; secretValue = $password } # Azure SQLServer
+              @{ name = 'vpnSharedKey'; secretValue = $vpnSharedKey } # VirtualNetworkGateway
+              @{ name = 'apimclientid'; secretValue = $username } # API management
+              @{ name = 'apimclientsecret'; secretValue = $password } # API management
+            ) | ForEach-Object {
+              $null = Set-AzKeyVaultSecret -VaultName $keyVaultName -Name $_.name -SecretValue $_.secretValue
+              Write-Verbose ('Added secret [{0}] to key vault [{1}]' -f $_.name, $keyVaultName) -Verbose
+            }
+
+            # Certificats
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:661]:
```suggestion
            # Certificates
```

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 1 above)
Comment (MrMCake) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep]: done

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 2 above)
Comment (MrMCake) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep]: done

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 1 above)
Comment (MrMCake) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep]: rolled back

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+                  @{ name = 'administratorLoginPassword'; secretValue = $password } # Azure SQLServer
+                  @{ name = 'vpnSharedKey'; secretValue = $vpnSharedKey } # VirtualNetworkGateway
+                  @{ name = 'apimclientid'; secretValue = $username } # API management
+                  @{ name = 'apimclientsecret'; secretValue = $password } # API management
+                ) | ForEach-Object {
+                  $null = Set-AzKeyVaultSecret -VaultName $keyVaultName -Name $_.name -SecretValue $_.secretValue
+                  Write-Verbose ('Added secret [{0}] to key vault [{1}]' -f $_.name, $keyVaultName) -Verbose
+                }
+
+                # Certificats
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:444]:
```suggestion
                # Certificates
```

File: .github/workflows/platform.dependencies.yml [hunk 8]
```diff
@@ -725,22 +726,39 @@ jobs:
...
-            $usernameString = (-join ((65..90) + (97..122) | Get-Random -Count 9 -SetSeed 1 | % {[char]$_ + "$_"})).substring(0,19) # max length
-            $passwordString = (New-Guid).Guid.SubString(0,19)
-            $vpnSharedKeyString =  (New-Guid).Guid.SubString(0,32)
-            $keyVaultName = 'adp-sxx-az-kv-x-sqlmi'
+            # Get key vault name
+            $parameterFilePath = Join-Path $env:GITHUB_WORKSPACE 'utilities' 'pipelines' 'dependencies' '${{ env.namespace }}' 'parameters' 'sqlmi.parameters.json'
+            $keyVaultParameters = (ConvertFrom-Json (Get-Content -Path $parameterFilePath -Raw)).parameters
+            $keyVaultName = $keyVaultParameters.name.value
+
+            # Prepare
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:738]:
```suggestion
            # Generate values
```

File: .github/workflows/platform.dependencies.yml [hunk 9]
```diff
@@ -725,22 +726,39 @@ jobs:
...
+            $passwordString = (New-Guid).Guid.SubString(0, 19)
             $password = ConvertTo-SecureString -String $passwordString -AsPlainText -Force
-            $vpnSharedKey = ConvertTo-SecureString -String $vpnSharedKeyString -AsPlainText -Force
-            # SQLManagedInstances secrets
-            Set-AzKeyVaultSecret -VaultName $keyVaultName -Name 'administratorLogin' -SecretValue $username
-            Set-AzKeyVaultSecret -VaultName $keyVaultName -Name 'administratorLoginPassword' -SecretValue $password
-            # SQLManagedInstances Keys
-            az keyvault key create --vault-name $keyVaultName --name 'keyEncryptionKeySqlMi'
+
+            # Secrets
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:744]: Set secrets

Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:744]:
```suggestion
            # Set secrets
```

File: .github/workflows/platform.dependencies.yml [hunk 10]
```diff
@@ -725,22 +726,39 @@ jobs:
...
+            # -------
+            @(
+              @{ name = 'administratorLogin'; secretValue = $username } # SQLManagedInstances
+              @{ name = 'administratorLoginPassword'; secretValue = $password } # SQLManagedInstances
+            ) | ForEach-Object {
+              $null = Set-AzKeyVaultSecret -VaultName $keyVaultName -Name $_.name -SecretValue $_.secretValue
+              Write-Verbose ('Added secret [{0}] to key vault [{1}]' -f $_.name, $keyVaultName) -Verbose
+            }
+
+            # Keys
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:754]:
```suggestion
            # Set keys
```

File: .github/workflows/platform.dependencies.yml [hunk 11]
```diff
@@ -650,33 +626,56 @@ jobs:
...
-            # DiskEncryptionSet, VirtualMachines and VMSS
-            az keyvault key create --vault-name $keyVaultName --name 'keyEncryptionKey'
-            # ApplicationGateway
-            $apgwCertPolicy = New-AzKeyVaultCertificatePolicy -SecretContentType 'application/x-pkcs12' -SubjectName 'CN=fabrikam.com' -IssuerName 'Self' -ValidityInMonths 12 -ReuseKeyOnRenewal
-            Add-AzKeyVaultCertificate -VaultName $keyVaultName -Name 'applicationGatewaySslCertificate' -CertificatePolicy $apgwCertPolicy
-            # API management
-            Set-AzKeyVaultSecret -VaultName $keyVaultName -Name 'apimclientid' -SecretValue $username
-            Set-AzKeyVaultSecret -VaultName $keyVaultName -Name 'apimclientsecret' -SecretValue $password
+
+            # Secrets
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:646]:
```suggestion
            # Set secrets
```

File: .github/workflows/platform.dependencies.yml [hunk 12]
```diff
@@ -650,33 +626,56 @@ jobs:
...
+            # -----------
+            $certPolicy = New-AzKeyVaultCertificatePolicy -SecretContentType 'application/x-pkcs12' -SubjectName 'CN=fabrikam.com' -IssuerName 'Self' -ValidityInMonths 12 -ReuseKeyOnRenewal
+            @(
+              @{ name = 'applicationGatewaySslCertificate'; CertificatePolicy = $certPolicy } # ApplicationGateway
+            ) | ForEach-Object {
+              $null = Add-AzKeyVaultCertificate -VaultName $keyVaultName -Name $_.name -CertificatePolicy $_.CertificatePolicy
+              Write-Verbose ('Added certificate [{0}] to key vault [{1}]' -f $_.name, $keyVaultName) -Verbose
+            }
+
+            # Keys
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:671]:
```suggestion
            # Set keys
```

File: .github/workflows/platform.dependencies.yml [hunk 13]
```diff
@@ -650,33 +626,56 @@ jobs:
...
-            $usernameString = (-join ((65..90) + (97..122) | Get-Random -Count 9 -SetSeed 1 | % {[char]$_ + "$_"})).substring(0,19) # max length
-            $passwordString = (New-Guid).Guid.SubString(0,19)
-            $vpnSharedKeyString =  (New-Guid).Guid.SubString(0,32)
-            $keyVaultName = 'adp-sxx-az-kv-x-001'
+            # Get key vault name
+            $parameterFilePath = Join-Path $env:GITHUB_WORKSPACE 'utilities' 'pipelines' 'dependencies' '${{ env.namespace }}' 'parameters' 'parameters.json'
+            $keyVaultParameters = (ConvertFrom-Json (Get-Content -Path $parameterFilePath -Raw)).parameters
+            $keyVaultName = $keyVaultParameters.name.value
+
+            # Prepare
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:638]:
```suggestion
            #  Generate values
```

File: .github/workflows/platform.dependencies.yml [hunk 14]
```diff
@@ -650,33 +626,56 @@ jobs:
...
+            # Secrets
+            # -------
+            @(
+              @{ name = 'adminUsername'; secretValue = $username } # VirtualMachines and VMSS
+              @{ name = 'adminPassword'; secretValue = $password } # VirtualMachines and VMSS
+              @{ name = 'administratorLogin'; secretValue = $username } # Azure SQLServer
+              @{ name = 'administratorLoginPassword'; secretValue = $password } # Azure SQLServer
+              @{ name = 'vpnSharedKey'; secretValue = $vpnSharedKey } # VirtualNetworkGateway
+              @{ name = 'apimclientid'; secretValue = $username } # API management
+              @{ name = 'apimclientsecret'; secretValue = $password } # API management
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:655]:
```suggestion
              @{ name = 'apimClientId'; secretValue = $username } # API management
              @{ name = 'apimClientSecret'; secretValue = $password } # API management
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 15]
```diff
@@ -0,0 +1,622 @@
...
+            inputs:
+              azureSubscription: $(serviceConnection)
+              ScriptType: 'InlineScript'
+              Inline: |
+                # Get key vault name
+                $parameterFilePath = Join-Path '$(Build.SourcesDirectory)' '$(dependencyPath)' '$(resourceType)' 'parameters' 'parameters.json'
+                $keyVaultParameters = (ConvertFrom-Json (Get-Content -Path $parameterFilePath -Raw)).parameters
+                $keyVaultName = $keyVaultParameters.name.value
+
+                # Prepare
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:421]:
```suggestion
                # Generate values
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 16]
```diff
@@ -0,0 +1,622 @@
...
+                $usernameString = ( -join ((65..90) + (97..122) | Get-Random -Count 9 -SetSeed 1 | ForEach-Object { [char]$_ + "$_" })).substring(0, 19) # max length
+                $userName = ConvertTo-SecureString -String $usernameString -AsPlainText -Force
+                $passwordString = (New-Guid).Guid.SubString(0, 19)
+                $password = ConvertTo-SecureString -String $passwordString -AsPlainText -Force
+                $vpnSharedKeyString = (New-Guid).Guid.SubString(0, 32)
+                $vpnSharedKey = ConvertTo-SecureString -String $vpnSharedKeyString -AsPlainText -Force
+
+                # Secrets
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:429]:
```suggestion
                # Set secrets
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 17]
```diff
@@ -0,0 +1,622 @@
...
+                # -----------
+                $certPolicy = New-AzKeyVaultCertificatePolicy -SecretContentType 'application/x-pkcs12' -SubjectName 'CN=fabrikam.com' -IssuerName 'Self' -ValidityInMonths 12 -ReuseKeyOnRenewal
+                @(
+                  @{ name = 'applicationGatewaySslCertificate'; CertificatePolicy = $certPolicy } # ApplicationGateway
+                ) | ForEach-Object {
+                  $null = Add-AzKeyVaultCertificate -VaultName $keyVaultName -Name $_.name -CertificatePolicy $_.CertificatePolicy
+                  Write-Verbose ('Added certificate [{0}] to key vault [{1}]' -f $_.name, $keyVaultName) -Verbose
+                }
+
+                # Keys
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:454]:
```suggestion
                # Set keys
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 18]
```diff
@@ -0,0 +1,622 @@
...
+                # Secrets
+                # -------
+                @(
+                  @{ name = 'adminUsername'; secretValue = $username } # VirtualMachines and VMSS
+                  @{ name = 'adminPassword'; secretValue = $password } # VirtualMachines and VMSS
+                  @{ name = 'administratorLogin'; secretValue = $username } # Azure SQLServer
+                  @{ name = 'administratorLoginPassword'; secretValue = $password } # Azure SQLServer
+                  @{ name = 'vpnSharedKey'; secretValue = $vpnSharedKey } # VirtualNetworkGateway
+                  @{ name = 'apimclientid'; secretValue = $username } # API management
+                  @{ name = 'apimclientsecret'; secretValue = $password } # API management
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:438]:
```suggestion
                  @{ name = 'apimClientId'; secretValue = $username } # API management
                  @{ name = 'apimClientSecret'; secretValue = $password } # API management
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 19]
```diff
@@ -0,0 +1,622 @@
...
+                inputs:
+                  azureSubscription: $(serviceConnection)
+                  ScriptType: 'InlineScript'
+                  Inline: |
+                    # Get key vault name
+                    $parameterFilePath = Join-Path '$(Build.SourcesDirectory)' '$(dependencyPath)' '$(resourceType)' 'parameters' 'sqlmi.parameters.json'
+                    $keyVaultParameters = (ConvertFrom-Json (Get-Content -Path $parameterFilePath -Raw)).parameters
+                    $keyVaultName = $keyVaultParameters.name.value
+
+                    # Prepare
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:499]:
```suggestion
                    # Generate values
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 20]
```diff
@@ -0,0 +1,622 @@
...
+                    $usernameString = ( -join ((65..90) + (97..122) | Get-Random -Count 9 -SetSeed 1 | ForEach-Object { [char]$_ + "$_" })).substring(0, 19) # max length
+                    $userName = ConvertTo-SecureString -String $usernameString -AsPlainText -Force
+                    $passwordString = (New-Guid).Guid.SubString(0, 19)
+                    $password = ConvertTo-SecureString -String $passwordString -AsPlainText -Force
+
+                    # Secrets
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:505]:
```suggestion
                    # Set secrets
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 21]
```diff
@@ -0,0 +1,622 @@
...
+                    # -------
+                    @(
+                      @{ name = 'administratorLogin'; secretValue = $username } # SQLManagedInstances
+                      @{ name = 'administratorLoginPassword'; secretValue = $password } # SQLManagedInstances
+                    ) | ForEach-Object {
+                      $null = Set-AzKeyVaultSecret -VaultName $keyVaultName -Name $_.name -SecretValue $_.secretValue
+                      Write-Verbose ('Added secret [{0}] to key vault [{1}]' -f $_.name, $keyVaultName) -Verbose
+                    }
+
+                    # Keys
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:515]:
```suggestion
                    # Set keys
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 22]
```diff
@@ -0,0 +1,622 @@
...
+            displayName: 'Setup agent'
+            inputs:
+              targetType: inline
+              pwsh: true
+              script: |
+                # Load used functions
+                . (Join-Path '$(System.DefaultWorkingDirectory)' 'utilities' 'pipelines' 'sharedScripts' 'Set-EnvironmentOnAgent.ps1')
+
+                # Set agent up
+                Set-EnvironmentOnAgent
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:150]:
```suggestion              
                # Define PS modules to install on the runner
                $Modules = @(
                    @{ Name = 'Az.Storage' }
                )

                # Set agent up
                Set-EnvironmentOnAgent -PSModules $Modules
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 23]
```diff
@@ -0,0 +1,622 @@
...
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:409]:
```suggestion
                # Define PS modules to install on the runner
                $Modules = @(
                    @{ Name = 'Az.KeyVault' }
                )

                # Set agent up
                Set-EnvironmentOnAgent -PSModules $Modules
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 24]
```diff
@@ -0,0 +1,622 @@
...
+                displayName: 'Setup agent'
+                inputs:
+                  targetType: inline
+                  pwsh: true
+                  script: |
+                    # Load used functions
+                    . (Join-Path '$(System.DefaultWorkingDirectory)' 'utilities' 'pipelines' 'sharedScripts' 'Set-EnvironmentOnAgent.ps1')
+
+                    # Set agent up
+                    Set-EnvironmentOnAgent
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:487]:
```suggestion
                # Define PS modules to install on the runner
                $Modules = @(
                    @{ Name = 'Az.KeyVault' }
                )

                # Set agent up
                Set-EnvironmentOnAgent -PSModules $Modules
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 25]
```diff
@@ -0,0 +1,622 @@
...
+        dependsOn:
+          - default_sa
+        pool:
+          ${{ if eq(variables['vmImage'], '') }}:
+            name: $(poolName)
+          ${{ if eq(variables['poolName'], '') }}:
+            vmImage: $(vmImage)
+        steps:
+          - task: PowerShell@2
+            displayName: 'Setup agent'
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:141]: Should we add the setup agent step in the gh workflow as well for kv and storage post deployment? 
Az.Storage and Az.KeyVault PS modules seem to be installed by default on agents as the pipeline does not fail, but for consistency I'd suggest so.

Comment (MrMCake) [reply .azuredevops/platformPipelines/platform.dependencies.yml:141]: Agreed. Will do

Comment (MrMCake) [reply .azuredevops/platformPipelines/platform.dependencies.yml:141]: done. Rerunning pipeline

Comment (MrMCake) [reply .azuredevops/platformPipelines/platform.dependencies.yml:141]: Success: https://github.com/Azure/ResourceModules/actions/runs/1687076011

**Review by eriqua — APPROVED** [review]
🆗

File: .github/workflows/platform.dependencies.yml (hunk 5 above)
Comment (eriqua) [reply .github/workflows/platform.dependencies.yml:41]: Fine for me

**Pull request merged**